File: csharp/src/Descriptor.cs
// The descriptor normalizer and canonical serializer (station design 4):
// a generated SDK's embedded config, normalized into descriptor v1 - a
// VIEW over the config, never a second model - plus the one canonical
// byte serialization every port must agree on.
//
// A port of typescript/src/descriptor.ts, which is canonical.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace Voxgig.Station
{
    public static class Descriptor
    {
        private static readonly Regex NonToken = new Regex("[^A-Z0-9]+", RegexOptions.Compiled);
        private static readonly Regex EdgeUnderscores = new Regex("^_+|_+$", RegexOptions.Compiled);

        /// <summary>A normalized descriptor plus any legacy warnings.</summary>
        public sealed class Normalized
        {
            public readonly Dictionary<string, object> Descriptor;
            public readonly List<string> Warnings;

            internal Normalized(Dictionary<string, object> descriptor, List<string> warnings)
            {
                Descriptor = descriptor;
                Warnings = warnings;
            }
        }

        /// <summary>
        /// The ONLY way to build an env-var token in station, mirroring
        /// sdkgen's packageMeta envToken exactly: 'gnarly-pets' ->
        /// 'GNARLY_PETS'. The `secretname` corpus section pins the round-trip
        /// against sekreto's EnvKey() and sdkgen's envName() - the one place
        /// three grammars meet.
        /// </summary>
        public static string EnvToken(object name)
        {
            string text = null == name ? "" : Convert.ToString(name, CultureInfo.InvariantCulture);
            return EdgeUnderscores.Replace(
                NonToken.Replace((text ?? "").ToUpperInvariant(), "_"), "");
        }

        /// <summary>
        /// The default sekreto name for a plugin (design 5.1): EnvToken(slug)
        /// lowercased, plus '.apikey'. sekreto's EnvKey() then yields exactly
        /// the env var the SDK's README documents: gnarly_pets.apikey ->
        /// GNARLY_PETS_APIKEY.
        /// </summary>
        public static string SecretnameDefault(string slug)
        {
            return EnvToken(slug).ToLowerInvariant() + ".apikey";
        }

        // Best-effort slug from a camel name, for SDKs whose embedded config
        // predates main.slug (design 4 legacy sentinels). The hyphen caveat is
        // real: 'VoxgigSolardemo' -> 'voxgigsolardemo', NOT 'voxgig-solardemo'
        // - callers surface a warning event when this path is taken.
        internal static string LegacySlug(string name)
        {
            return (null == name ? "" : name).ToLowerInvariant();
        }

        /// <summary>
        /// Normalize a generated SDK's embedded config into descriptor v1
        /// (design 4). The config is the one every SDK carries (Config.main /
        /// .feature / .options / .entity); the descriptor is a VIEW over it.
        /// Returns the descriptor plus any legacy warnings.
        /// </summary>
        public static Normalized NormalizeDescriptor(object config, object activeFeatures)
        {
            var warnings = new List<string>();
            Dictionary<string, object> main = AsMap(GetProp(config, "main"));
            Dictionary<string, object> options = AsMap(GetProp(config, "options"));

            string name = TextOr(main.GetValueOrDefault("name"), "");
            string slug = TextOr(main.GetValueOrDefault("slug"), "");
            if (0 == slug.Length)
            {
                slug = LegacySlug(name);
                warnings.Add("descriptor: legacy config has no main.slug; derived \"" + slug
                    + "\" from the camel name - hyphens in the original name are lost");
            }

            string version = null == main.GetValueOrDefault("version")
                ? "0.0.0" : Convert.ToString(main["version"], CultureInfo.InvariantCulture);
            string target = null == main.GetValueOrDefault("target")
                ? "unknown" : Convert.ToString(main["target"], CultureInfo.InvariantCulture);

            var server = new List<object>();
            Dictionary<string, object> svr = AsMap(options.GetValueOrDefault("server"));
            foreach (string key in SortedKeys(svr))
            {
                server.Add(new Dictionary<string, object>
                {
                    ["name"] = key,
                    ["value"] = Convert.ToString(svr[key], CultureInfo.InvariantCulture),
                });
            }

            bool authActive = null != options.GetValueOrDefault("auth");
            var auth = new Dictionary<string, object>
            {
                ["active"] = authActive,
                ["prefix"] = authActive
                    ? TextOr(GetProp(options["auth"], "prefix"), "") : "",
                ["secretname"] = SecretnameDefault(slug),
            };

            var entities = new Dictionary<string, object>();
            Dictionary<string, object> entdefs = AsMap(GetProp(config, "entity"));
            foreach (string ename in SortedKeys(entdefs))
            {
                Dictionary<string, object> e = AsMap(entdefs[ename]);

                var fields = new Dictionary<string, object>();
                if (e.GetValueOrDefault("fields") is IList<object> fdefs)
                {
                    foreach (object f in fdefs)
                    {
                        object fname = GetProp(f, "name");
                        if (null != f && null != fname)
                        {
                            object kind = GetProp(f, "kind");
                            if (null == kind)
                            {
                                kind = GetProp(f, "type");
                            }
                            fields[Convert.ToString(fname, CultureInfo.InvariantCulture)] =
                                new Dictionary<string, object>
                                {
                                    ["kind"] = null == kind
                                        ? "" : Convert.ToString(kind, CultureInfo.InvariantCulture),
                                };
                        }
                    }
                }

                var ops = new Dictionary<string, object>();
                Dictionary<string, object> opdefs = AsMap(e.GetValueOrDefault("op"));
                foreach (string opname in SortedKeys(opdefs))
                {
                    Dictionary<string, object> op = AsMap(opdefs[opname]);
                    var points = new List<object>();
                    if (op.GetValueOrDefault("points") is IList<object> pdefs)
                    {
                        foreach (object p in pdefs)
                        {
                            if (null == p)
                            {
                                continue;
                            }
                            var point = new Dictionary<string, object>
                            {
                                ["method"] = TextOr(GetProp(p, "method"), ""),
                            };
                            object path = GetProp(p, "orig");
                            if (null == path)
                            {
                                path = GetProp(p, "path");
                            }
                            point["path"] = null == path
                                ? "" : Convert.ToString(path, CultureInfo.InvariantCulture);

                            var pparams = new List<object>();
                            if (GetProp(p, "parts") is IList<object> parts)
                            {
                                foreach (object part in parts)
                                {
                                    if (part is string ptext && ptext.StartsWith(":"))
                                    {
                                        pparams.Add(ptext.Substring(1));
                                    }
                                }
                            }
                            point["params"] = pparams;

                            object select = GetProp(p, "select");
                            if (null != select)
                            {
                                point["select"] = select;
                            }
                            points.Add(point);
                        }
                    }
                    ops[opname] = new Dictionary<string, object> { ["points"] = points };
                }

                entities[ename] = new Dictionary<string, object>
                {
                    ["fields"] = fields,
                    ["ops"] = ops,
                };
            }

            var features = new List<object>();
            Dictionary<string, object> featdefs = AsMap(GetProp(config, "feature"));
            Dictionary<string, object> factive = AsMap(activeFeatures);
            foreach (string fname in SortedKeys(featdefs))
            {
                features.Add(new Dictionary<string, object>
                {
                    ["name"] = fname,
                    ["active"] = true.Equals(GetProp(factive.GetValueOrDefault(fname), "active")),
                });
            }

            var descriptor = new Dictionary<string, object>
            {
                ["station"] = 1,
                ["name"] = name,
                ["slug"] = slug,
                ["envtoken"] = EnvToken(slug),
                ["version"] = version,
                ["target"] = target,
                ["base"] = TextOr(options.GetValueOrDefault("base"), ""),
                ["server"] = server,
                ["auth"] = auth,
                ["entities"] = entities,
                ["features"] = features,
            };

            return new Normalized(descriptor, warnings);
        }

        /// <summary>
        /// Canonical serialization (design 4): UTF-8, object keys sorted
        /// bytewise, no insignificant whitespace, minimal JSON escaping. The
        /// proxy dedupes registrations by a hash of this, so every language
        /// must produce identical bytes - the `canonical-serialize` corpus
        /// section carries the adversarial cases.
        /// </summary>
        public static string CanonicalSerialize(object value)
        {
            var out_ = new StringBuilder();
            Serialize(value, out_);
            return out_.ToString();
        }

        private static void Serialize(object value, StringBuilder out_)
        {
            if (null == value)
            {
                out_.Append("null");
                return;
            }
            if (value is bool flag)
            {
                out_.Append(flag ? "true" : "false");
                return;
            }
            if (value is string text)
            {
                Quote(text, out_);
                return;
            }
            if (IsNumber(value))
            {
                out_.Append(NumStr(value));
                return;
            }
            if (value is IDictionary<string, object> map)
            {
                var keys = new List<string>(map.Keys);
                // Bytewise sort: compare UTF-8 byte sequences, not UTF-16
                // code units.
                keys.Sort(CompareUtf8);
                out_.Append('{');
                bool first = true;
                foreach (string key in keys)
                {
                    if (!first)
                    {
                        out_.Append(',');
                    }
                    first = false;
                    Quote(key, out_);
                    out_.Append(':');
                    Serialize(map[key], out_);
                }
                out_.Append('}');
                return;
            }
            if (value is IList<object> list)
            {
                out_.Append('[');
                for (int index = 0; index < list.Count; index++)
                {
                    if (0 < index)
                    {
                        out_.Append(',');
                    }
                    Serialize(list[index], out_);
                }
                out_.Append(']');
                return;
            }
            out_.Append("null");
        }

        internal static bool IsNumber(object value)
        {
            return value is int || value is long || value is double
                || value is float || value is short || value is byte
                || value is decimal || value is uint || value is ulong
                || value is ushort || value is sbyte;
        }

        // Integral numbers print without a decimal point or exponent, matching
        // the canonical JSON.stringify behaviour (2^53-1 stays exact).
        internal static string NumStr(object value)
        {
            double d = Convert.ToDouble(value, CultureInfo.InvariantCulture);
            if (!double.IsInfinity(d) && d == Math.Floor(d)
                && long.MinValue <= d && d <= long.MaxValue)
            {
                return ((long)d).ToString(CultureInfo.InvariantCulture);
            }
            return d.ToString("R", CultureInfo.InvariantCulture);
        }

        // Minimal JSON escaping: quote, backslash, and control characters
        // only; everything else (non-ASCII included) passes through as UTF-8.
        internal static void Quote(string text, StringBuilder out_)
        {
            out_.Append('"');
            foreach (char c in text)
            {
                switch (c)
                {
                    case '"':
                        out_.Append("\\\"");
                        break;
                    case '\\':
                        out_.Append("\\\\");
                        break;
                    case '\b':
                        out_.Append("\\b");
                        break;
                    case '\f':
                        out_.Append("\\f");
                        break;
                    case '\n':
                        out_.Append("\\n");
                        break;
                    case '\r':
                        out_.Append("\\r");
                        break;
                    case '\t':
                        out_.Append("\\t");
                        break;
                    default:
                        if (c < 0x20)
                        {
                            out_.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                        }
                        else
                        {
                            out_.Append(c);
                        }
                        break;
                }
            }
            out_.Append('"');
        }

        internal static int CompareUtf8(string a, string b)
        {
            // C# bytes are unsigned, so a UTF-8 lead byte (0x80+) already
            // sorts AFTER ASCII without any masking.
            byte[] ab = Encoding.UTF8.GetBytes(a);
            byte[] bb = Encoding.UTF8.GetBytes(b);
            int n = Math.Min(ab.Length, bb.Length);
            for (int index = 0; index < n; index++)
            {
                int c = ab[index].CompareTo(bb[index]);
                if (0 != c)
                {
                    return c;
                }
            }
            return ab.Length.CompareTo(bb.Length);
        }

        // --- small shared helpers (map-shaped data, sekreto style) ---

        public static Dictionary<string, object> AsMap(object val)
        {
            if (val is Dictionary<string, object> map)
            {
                return map;
            }
            if (val is IDictionary<string, object> imap)
            {
                return new Dictionary<string, object>(imap);
            }
            return new Dictionary<string, object>();
        }

        public static object GetProp(object val, string key)
        {
            if (val is IDictionary<string, object> map)
            {
                return map.TryGetValue(key, out object found) ? found : null;
            }
            return null;
        }

        internal static string TextOr(object val, string alt)
        {
            if (val is string text && 0 != text.Length)
            {
                return text;
            }
            return alt;
        }

        private static List<string> SortedKeys(Dictionary<string, object> map)
        {
            var keys = new List<string>(map.Keys);
            keys.Sort(StringComparer.Ordinal);
            return keys;
        }
    }
}

File: dart/lib/src/descriptor.dart
// Descriptor v1 (design station.md 4): a view over the SDK's embedded
// config, normalized - never a second model. Plus the canonical
// serializer and the envtoken/secretname grammar.
//
// A port of typescript/src/descriptor.ts, which is canonical.

import 'dart:convert';

// The ONLY way to build an env-var token in station, mirroring sdkgen's
// packageMeta envToken exactly: 'gnarly-pets' -> 'GNARLY_PETS'. The
// `secretname` corpus section pins the round-trip against sekreto's
// envkey() and sdkgen's envName() - the one place three grammars meet.
String envtoken(dynamic name) {
  return (name ?? '')
      .toString()
      .toUpperCase()
      .replaceAll(RegExp(r'[^A-Z0-9]+'), '_')
      .replaceAll(RegExp(r'^_+|_+$'), '');
}

// The default sekreto name for a plugin (design station.md 5.1):
// envtoken(slug) lowercased, plus '.apikey'. sekreto's envkey() then
// yields exactly the env var the SDK's README documents:
// gnarly_pets.apikey -> GNARLY_PETS_APIKEY.
String secretnameDefault(dynamic slug) {
  return envtoken(slug).toLowerCase() + '.apikey';
}

// Best-effort slug from a camel name, for SDKs whose embedded config
// predates main.slug (design station.md 4 legacy sentinels). The hyphen
// caveat is real: 'VoxgigSolardemo' -> 'voxgigsolardemo', NOT
// 'voxgig-solardemo' - callers surface a warning event when this path is
// taken.
String _legacySlug(dynamic name) {
  return (name ?? '').toString().toLowerCase();
}

// The normalizer's result: the descriptor (a plain string-keyed map, the
// shape the corpus pins) plus any legacy warnings.
class Normalized {
  final Map<String, dynamic> descriptor;
  final List<String> warnings;
  Normalized(this.descriptor, this.warnings);
}

dynamic _mget(dynamic m, String k) => m is Map ? m[k] : null;

// Normalize a generated SDK's embedded config into descriptor v1
// (design station.md 4). The config is the one every SDK carries
// (Config.main / .feature / .options / .entity); the descriptor is a
// VIEW over it.
Normalized normalizeDescriptor(dynamic config, [dynamic activeFeatures]) {
  final warnings = <String>[];
  final main = _mget(config, 'main') ?? {};
  final options = _mget(config, 'options') ?? {};

  final name = (_mget(main, 'name') ?? '').toString();
  var slug = _mget(main, 'slug');
  if (null == slug || '' == slug) {
    slug = _legacySlug(name);
    warnings.add('descriptor: legacy config has no main.slug; derived "' +
        slug.toString() +
        '" from the camel name - hyphens in the original name are lost');
  }
  slug = slug.toString();

  final version = null != _mget(main, 'version')
      ? _mget(main, 'version').toString()
      : '0.0.0';
  final target = null != _mget(main, 'target')
      ? _mget(main, 'target').toString()
      : 'unknown';

  final server = <Map<String, dynamic>>[];
  final svr = _mget(options, 'server') ?? {};
  if (svr is Map) {
    final skeys = svr.keys.map((k) => k.toString()).toList()..sort();
    for (final k in skeys) {
      server.add(<String, dynamic>{'name': k, 'value': svr[k].toString()});
    }
  }

  final authActive = null != _mget(options, 'auth');
  final auth = <String, dynamic>{
    'active': authActive,
    'prefix':
        authActive ? (_mget(_mget(options, 'auth'), 'prefix') ?? '').toString() : '',
    'secretname': secretnameDefault(slug),
  };

  final entities = <String, dynamic>{};
  final entdefs = _mget(config, 'entity') ?? {};
  if (entdefs is Map) {
    final enames = entdefs.keys.map((k) => k.toString()).toList()..sort();
    for (final ename in enames) {
      final e = entdefs[ename] ?? {};
      final fields = <String, dynamic>{};
      final flist = _mget(e, 'fields');
      if (flist is List) {
        for (final f in flist) {
          if (f is Map && null != f['name']) {
            var kind = f['kind'];
            if (null == kind || '' == kind) {
              kind = f['type'];
            }
            fields[f['name'].toString()] = <String, dynamic>{
              'kind': (kind ?? '').toString(),
            };
          }
        }
      }
      final ops = <String, dynamic>{};
      final opdefs = _mget(e, 'op') ?? {};
      if (opdefs is Map) {
        final opnames = opdefs.keys.map((k) => k.toString()).toList()..sort();
        for (final opname in opnames) {
          final op = opdefs[opname] ?? {};
          final points = <Map<String, dynamic>>[];
          final plist = _mget(op, 'points');
          if (plist is List) {
            for (final p in plist) {
              if (p is! Map) {
                continue;
              }
              final params = <String>[];
              final parts = p['parts'];
              if (parts is List) {
                for (final s in parts) {
                  if (s is String && s.startsWith(':')) {
                    params.add(s.substring(1));
                  }
                }
              }
              final point = <String, dynamic>{
                'method': (p['method'] ?? '').toString(),
                'path': (p['orig'] ?? p['path'] ?? '').toString(),
                'params': params,
              };
              if (null != p['select']) {
                point['select'] = p['select'];
              }
              points.add(point);
            }
          }
          ops[opname] = <String, dynamic>{'points': points};
        }
      }
      entities[ename] = <String, dynamic>{'fields': fields, 'ops': ops};
    }
  }

  final features = <Map<String, dynamic>>[];
  final fdefs = _mget(config, 'feature') ?? {};
  final factive = activeFeatures is Map ? activeFeatures : {};
  if (fdefs is Map) {
    final fnames = fdefs.keys.map((k) => k.toString()).toList()..sort();
    for (final fname in fnames) {
      features.add(<String, dynamic>{
        'name': fname,
        'active': true == _mget(factive[fname], 'active'),
      });
    }
  }

  final descriptor = <String, dynamic>{
    'station': 1,
    'name': name,
    'slug': slug,
    'envtoken': envtoken(slug),
    'version': version,
    'target': target,
    'base': (_mget(options, 'base') ?? '').toString(),
    'server': server,
    'auth': auth,
    'entities': entities,
    'features': features,
  };

  return Normalized(descriptor, warnings);
}

// Canonical serialization (design station.md 4): UTF-8, object keys
// sorted bytewise, no insignificant whitespace, minimal JSON escaping.
// The proxy dedupes registrations by a hash of this, so every language
// must produce identical bytes - the `canonical` corpus section carries
// the adversarial cases.
String canonicalSerialize(dynamic value) {
  if (null == value) {
    return 'null';
  }
  if (value is bool) {
    return value ? 'true' : 'false';
  }
  if (value is int) {
    return value.toString();
  }
  if (value is num) {
    // A whole double serializes as the integer it is, matching the
    // reference JSON.stringify - the descriptor's numeric fields are
    // integers by definition.
    final d = value.toDouble();
    if (d.isFinite && d == d.truncateToDouble() && d.abs() < 9007199254740992.0) {
      return d.truncate().toString();
    }
    return value.toString();
  }
  if (value is String) {
    return _jsonString(value);
  }
  if (value is List) {
    return '[' + value.map(canonicalSerialize).join(',') + ']';
  }
  if (value is Map) {
    final keys = value.keys.map((k) => k.toString()).toList();
    keys.sort(_bytewise);
    return '{' +
        keys
            .map((k) => _jsonString(k) + ':' + canonicalSerialize(value[k]))
            .join(',') +
        '}';
  }
  return 'null';
}

// Bytewise sort: compare UTF-8 byte sequences, not UTF-16 code units.
int _bytewise(String a, String b) {
  final ab = utf8.encode(a);
  final bb = utf8.encode(b);
  final n = ab.length < bb.length ? ab.length : bb.length;
  for (var i = 0; i < n; i++) {
    if (ab[i] != bb[i]) {
      return ab[i] - bb[i];
    }
  }
  return ab.length - bb.length;
}

// Minimal JSON string escaping, matching the reference JSON.stringify:
// quote and backslash, the short escapes, \u00xx for other control
// characters; everything else (non-ASCII included) passes through raw.
String _jsonString(String s) {
  final out = StringBuffer('"');
  for (final unit in s.codeUnits) {
    switch (unit) {
      case 0x22:
        out.write(r'\"');
        break;
      case 0x5c:
        out.write(r'\\');
        break;
      case 0x08:
        out.write(r'\b');
        break;
      case 0x09:
        out.write(r'\t');
        break;
      case 0x0a:
        out.write(r'\n');
        break;
      case 0x0c:
        out.write(r'\f');
        break;
      case 0x0d:
        out.write(r'\r');
        break;
      default:
        if (unit < 0x20) {
          out.write('\\u' + unit.toRadixString(16).padLeft(4, '0'));
        } else {
          out.writeCharCode(unit);
        }
    }
  }
  out.write('"');
  return out.toString();
}
